src/attributes.jl

```diff
@@ -1856,6 +1856,108 @@ function attribute_value_type(::ConstraintConflictStatus)
     return ConflictParticipationStatusCode
 end
 
+"""
+    UserDefinedFunction(name::Symbol, arity::Int) <: AbstractModelAttribute
+
+Set this attribute to register a user-defined function by the name of `name`
+with `arity` arguments.
+
+Once registered, `name` will appear in
+[`ListOfSupportedNonlinearOperators`](@ref).
+
+## Value type
+
+The value to be set is a tuple containing one, two, or three functions to
+evaluate the function, the first-order derivative, and the second-order
+derivative respectively. Both derivatives are optional, and if you pass the
+second-order derivative you must also pass the first-order derivative.
+
+For univariate functions with `arity == 1`, the functions in the tuple must
+have the form:
+
+ * `f(x::T)::T`: returns the value of the function at `x`
+ * `∇f(x::T)::T`: returns the first-order derivative of `f` with respect to `x`
+ * `∇²f(x::T)::T`: returns the second-order derivative of `f` with respect to
+   `x`.
+
+For multivariate functions with `arity > 1`, the functions in the tuple must
+have the form:
+
+ * `f(x::T...)::T`: returns the value of the function at `x`
+ * `∇f(g::AbstractVector{T}, x::T...)::Nothing`: fills the components of `g`,
+   with `g[i]` being the first-order partial derivative of `f` with respect to
+   `x[i]`
+ * `∇²f(H::AbstractMatrix{T}, x::T...)::Nothing`: fills the components of `H`,
+    with `H[i, j]` being the second-order partial derivative of `f` with respect
+    to `x[i]` and then `x[j]`. It does not need to fill in the structural
```

Comment:
This is not clear, what is the difference between a structural and non-structural zero ? Can't we just say that `H` can be assumed to be a zero matrix at first so whatever is not set is just zero ?

Reply:
A structural zero is an element that is always zero. A non-structural zero is an element that is zero right now, but may be non-zero at a different evaluation point.

But yeah, too complicated. Assumed to be zero works better :+1: